tensorboard/plugins/profile/pod_viewer/pod_viewer_dashboard/pod-viewer-dashboard.ts

```diff
@@ -0,0 +1,235 @@
+  _computePodStatsMaps(data: podviewer.proto.PodViewerInputData|undefined|null):
+                       Array<podviewer.proto.PodStatsMap> {
+    if (!data) return [];
+    return data.podStatsSequence.podStatsMap;
+  },
+  _computeRunEnvironment(
+      data: podviewer.proto.PodViewerInputData|undefined|null):
+          podviewer.proto.RunEnvironment {
+    return data.runEnvironment;
+  },
+  _computeMaxStepId(podStatsMaps: Array<podviewer.proto.PodStatsMap>): number {
+    return podStatsMaps.length - 1;
+  },
+  _computeErrorMessage(maxStepId: number): string {
+    if (maxStepId >= 0) return '';
+    return "WARNING: No step time measured. "
+           + "This might happen if your profile duration is too short, "
+           + "try increase profile duration to cover a full step. "
+           + "If you have an inference job or not use TpuEstimator, "
+           + "please skip this tool.";
+  },
+  /**
+   * Calculate the lowFlopsComputeUs by deducting all other breakdown from the
+   * total duration.
+   */
+  _populateLowFlopsCompute(podStatsMap: podviewer.proto.PodStatsMap|undefined,
+                           layers: Array<podviewer.proto.StackLayer>):
+                           podviewer.proto.PodStatsMap {
+    if (!podStatsMap || !layers) return;
+    let podStatsPerCore = podStatsMap.podStatsPerCore;
+    for (let coreId in podStatsPerCore) {
+      let val = podStatsPerCore[coreId];
+      if (val.hasOwnProperty('lowFlopsComputeUs')) {
+        // already populated.
+        return podStatsMap;
+      }
+      val['lowFlopsComputeUs'] = val.totalDurationUs;
+      for (let j = 0; j < layers.length; j++) {
+        if (j == 1) {
+          continue;
+        }
+        // Skip the lowFlopsComputeUs.
+        val['lowFlopsComputeUs'] -= val[layers[j].key];
+      }
+    }
+   return podStatsMap;
+  },
+  _computePodStatsMap(podStatsMaps: Array<podviewer.proto.PodStatsMap>,
+                      curStepId: number,
+                      layers: Array<podviewer.proto.StackLayer>):
+                      podviewer.proto.PodStatsMap {
+    if (curStepId < 0 || curStepId >= podStatsMaps.length || !layers) return;
+    return this._populateLowFlopsCompute(podStatsMaps[curStepId], layers);
+  },
+  _computeStepStats(podStatsMap: podviewer.proto.PodStatsMap):
+                    Array<podviewer.proto.PodStatsRecord>|undefined {
+    if (!podStatsMap || !podStatsMap.podStatsPerCore) return;
+    let stepStats = [];
+    for (const i in podStatsMap.podStatsPerCore) {
+      stepStats.push(podStatsMap.podStatsPerCore[i]);
+    }
+    stepStats.sort((a, b) => a.chipId - b.chipId);
+    return stepStats;
+  },
+  _computeChannelDb(podStatsMap: podviewer.proto.PodStatsMap):
+                    Array<podviewer.proto.ChannelInfo>|undefined {
+    if (!podStatsMap || !podStatsMap.channelDb
+        || podStatsMap.channelDb.length <= 0) {
+      return;
+    }
+    return podStatsMap.channelDb.sort((a, b) => b.durationUs - a.durationUs);
```

Comment:
Just curious, what's the reason for no clone here? `sort` mutates the original array.
Here and below.

Reply:
Yes, I think .slice() is better. Thanks.